File: src/event.js
let events = [];

function event(name, data) {
  events.push({ name, data: JSON.stringify(data) });
}

function list() {
  if (events.length) {
    return events;
  }
}

function clear() {
  events = [];
}

module.exports.event = event;
module.exports.list = list;
module.exports.clear = clear;

File: rust/src/event.rs
use serde::{Deserialize, Serialize};

/// Mirrors `src/event.js`: a user statement can call `event(name, data)`,
/// and every emitted event since the last drain rides along with that
/// statement's result.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EventRecord {
    pub name: String,
    /// JSON-encoded payload, matching the original's `JSON.stringify(data)`.
    pub data: String,
}
